### PR TITLE
feat: add "Face ID" login method

### PR DESCRIPTION
**Pull request by HGZ-20**
Fix: https://github.com/casdoor/casdoor/issues/2537

Face Login via face-api.js

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "face-api.js": "^0.22.2"
```
Comment (hsluoyz) [inline package.json:3]: Why is here in root folder?

File: web/public/models/face_landmark_68_model-weights_manifest.json [hunk 2]
```diff
@@ -0,0 +1 @@
...pe":"uint8","scale":0.006002906724518421,"min":-0.7923836876364315}},{"name":"dense3/conv2/pointwise_filter","shape":[1,1,256,256],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.017087187019048954,"min":-1.6061955797906016}},{"name":"dense3/conv2/bias","shape":[256],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.003124481205846749,"min":-0.46242321846531886}},{"name":"dense3/conv3/depthwise_filter","shape":[3,3,256,1],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.006576311588287353,"min":-1.0193282961845398}},{"name":"dense3/conv3/pointwise_filter","shape":[1,1,256,256],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.015590153955945782,"min":-1.99553970636106}},{"name":"dense3/conv3/bias","shape":[256],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.004453541601405424,"min":-0.6546706154065973}},{"name":"fc/weights","shape":[256,136],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.010417488509533453,"min":-1.500118345372817}},{"name":"fc/bias","shape":[136],"dtype":"float32","quantization":{"dtype":"uint8","scale":0.0025084222648658005,"min":0.07683877646923065}}],"paths":["face_landmark_68_model-shard1"]}]
```
Comment (hsluoyz) [inline web/public/models/face_landmark_68_model-weights_manifest.json:1]: Is this model data?

File: web/src/ApplicationEditPage.js [hunk 3]
```diff
@@ -1049,7 +1049,7 @@ class ApplicationEditPage extends React.Component {
   submitApplicationEdit(exitAfterSave) {
     const application = Setting.deepCopy(this.state.application);
     application.providers = application.providers?.filter(provider => this.state.providers.map(provider => provider.name).includes(provider.name));
-    application.signinMethods = application.signinMethods?.filter(signinMethod => ["Password", "Verification code", "WebAuthn", "LDAP"].includes(signinMethod.name));
+    application.signinMethods = application.signinMethods?.filter(signinMethod => ["Password", "Verification code", "WebAuthn", "LDAP", "Face"].includes(signinMethod.name));
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:1052]: Face -> Face ID

File: web/src/auth/LoginPage.js [hunk 4]
```diff
@@ -999,6 +1003,19 @@ class LoginPage extends React.Component {
           </div>
         </Col>
       );
+    } else if (this.state.loginMethod === "face") {
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:1006]: face -> faceId

File: web/src/auth/LoginPage.js [hunk 5]
```diff
@@ -1020,6 +1037,7 @@ class LoginPage extends React.Component {
       [generateItemKey("Verification code", "Phone only"), {label: i18next.t("login:Verification code"), key: "verificationCodePhone"}],
       [generateItemKey("WebAuthn", "None"), {label: i18next.t("login:WebAuthn"), key: "webAuthn"}],
       [generateItemKey("LDAP", "None"), {label: i18next.t("login:LDAP"), key: "ldap"}],
+      [generateItemKey("Face", "None"), {label: i18next.t("login:Face"), key: "face"}],
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:1040]: Face -> Face ID

face -> faceId

File: web/src/table/SigninMethodTable.js [hunk 6]
```diff
@@ -71,6 +71,7 @@ class SigninMethodTable extends React.Component {
       {name: "Verification code", displayName: i18next.t("login:Verification code")},
       {name: "WebAuthn", displayName: i18next.t("login:WebAuthn")},
       {name: "LDAP", displayName: i18next.t("login:LDAP")},
+      {name: "Face", displayName: i18next.t("login:Face")},
```
Comment (hsluoyz) [inline web/src/table/SigninMethodTable.js:74]: Face ID

File: object/user.go [hunk 7]
```diff
@@ -665,7 +666,7 @@ func UpdateUser(id string, user *User, columns []string, isAdmin bool) (bool, er
 		columns = []string{
 			"owner", "display_name", "avatar", "first_name", "last_name",
 			"location", "address", "country_code", "region", "language", "affiliation", "title", "id_card_type", "id_card", "homepage", "bio", "tag", "language", "gender", "birthday", "education", "score", "karma", "ranking", "signup_application",
-			"is_admin", "is_forbidden", "is_deleted", "hash", "is_default_avatar", "properties", "webauthnCredentials", "managedAccounts",
+			"is_admin", "is_forbidden", "is_deleted", "hash", "is_default_avatar", "properties", "webauthnCredentials", "managedAccounts", "face_i_d",
```
Comment (hsluoyz) [inline object/user.go:669]: face_i_d -> faceId

File: object/user.go [hunk 8]
```diff
@@ -190,6 +190,7 @@ type User struct {
 	MultiFactorAuths    []*MfaProps           `xorm:"-" json:"multiFactorAuths,omitempty"`
 	Invitation          string                `xorm:"varchar(100) index" json:"invitation"`
 	InvitationCode      string                `xorm:"varchar(100) index" json:"invitationCode"`
+	FaceID              []float64             `json:"faceID"`
```
Comment (hsluoyz) [inline object/user.go:193]: FaceId

File: object/verification.go [hunk 9]
```diff
@@ -236,6 +237,24 @@ func CheckSigninCode(user *User, dest, code, lang string) error {
 	}
 }
 
+func CheckFace(user *User, faceID []float64, lang string) error {
```
Comment (hsluoyz) [inline object/verification.go:240]: CheckFaceId

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (HGZ-20) [reply package.json]: Already removed

File: web/public/models/face_landmark_68_model-weights_manifest.json (hunk 2 above)
Comment (HGZ-20) [reply web/public/models/face_landmark_68_model-weights_manifest.json]: Yes, currently imported via web resources, subsequently replaced with casbin's own cdn

File: web/yarn.lock [hunk 10]
```diff
@@ -5600,9 +5632,9 @@ caniuse-api@^3.0.0:
     lodash.uniq "^4.5.0"
 
 caniuse-lite@^1.0.0, caniuse-lite@^1.0.30001464, caniuse-lite@^1.0.30001503:
-  version "1.0.30001507"
-  resolved "https://registry.yarnpkg.com/caniuse-lite/-/caniuse-lite-1.0.30001507.tgz#fae53f6286e7564783eadea9b447819410a59534"
-  integrity sha512-SFpUDoSLCaE5XYL2jfqe9ova/pbQHEmbheDf5r4diNwbAgR3qxM9NQtfsiSscjqoya5K7kFcHPUQ+VsUkIJR4A==
+  version "1.0.30001597"
+  resolved "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001597.tgz"
```
Comment (hsluoyz) [inline web/yarn.lock:5636]: Use: yarnpkg.com

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request merged**